First iteration of air_quality API based on the temperature API

RaresCon opened the pull request:
I've written the base of `air_quality` API. I will add documentation as I progress with this PR. I would be glad to receive feedback, mainly on `read_data_sync`.

RaresCon committed:
First iteration of air_quality API

I've written the base of air_quality API, next I will add a fake sensor for
testing the API, including unit testing for the fake sensor

alexandruradovici commented:
Looks good, a few comments:
- rebase your branch to the updated master
- export the air_quality module in `lib.rs`

alexandruradovici commented on apis/air_quality/src/lib.rs:
```diff
@@ -0,0 +1,90 @@
+    }
+}
+
+pub struct AirQualityListener<F: Fn(i32)>(pub F);
```
You don't have to add a new listener, you can use a Cell<(i32,)> (a tuple with one element) that already implements `Upcall`. An example of usage is this driver https://github.com/UPB-CS-OpenSourceUpstream/libtock-rs/blob/4719e4e5c0dbd61715eda10b248f779a1622f57c/apis/console/src/lib.rs#L42

RaresCon committed:
Modified the API to include the suggested changes

I've removed the `AirQualityListener` and used a `Cell<Option<(u32,)>>`
instead. I've added a private Enum for choosing what kind of data reading
to execute, a private method that uses this Enum and two other public
functions as wrappers.

RaresCon committed:
Added fake `AirQuality` driver

I've added a fake `AirQuality` driver and some integration tests for it.
Every test is passed and should reflect the behaviour of the real driver.

RaresCon committed:
Added unittests for AirQuality API

I've added unittests for each function of the AirQuality API, all pass
as expected. Documentation for the API will be part of the next commit.

alexandruradovici commented on apis/air_quality/src/lib.rs:
```diff
@@ -0,0 +1,92 @@
+#[cfg(test)]
+mod tests;
+
+enum ReadType {
```
```suggestion
enum Value {
```

alexandruradovici commented on apis/air_quality/src/lib.rs:
```diff
@@ -0,0 +1,92 @@
+use libtock_platform::{
+    share::scope, share::Handle, DefaultConfig, ErrorCode, Subscribe, Syscalls,
+};
+use ReadType::{ReadCO2, ReadTVOC};
```
```suggestion
use Value::{ReadCO2, ReadTVOC};
```

alexandruradovici commented on apis/air_quality/src/lib.rs:
```diff
@@ -0,0 +1,92 @@
+mod tests;
+
+enum ReadType {
+    ReadCO2 = READ_CO2 as isize,
+    ReadTVOC = READ_TVOC as isize,
```
```suggestion
    CO2 = READ_CO2 as isize,
    TVOC = READ_TVOC as isize,
```

RaresCon replied:
I renamed the `ReadTVOC` value to `Tvoc` to avoid a warning. If needed, I will rename it to `TVOC`.

alexandruradovici commented on apis/air_quality/src/lib.rs:
```diff
@@ -0,0 +1,92 @@
+        Self::read_data_sync(ReadTVOC)
+    }
+
+    fn read_data_sync(read_type: ReadType) -> Result<u32, ErrorCode> {
```
Add a `read_sync` function that reads boths and returns a tuple.

RaresCon replied:
Added the function, I am using a `Cell<Option<(u32, u32)>>` in the fake driver. I will try to find a simpler way of scheduling two upcalls one after the other.

alexandruradovici commented on apis/air_quality/src/tests.rs:
```diff
@@ -0,0 +1,106 @@
+}
+
+#[test]
+fn read_co2() {
```
Verify that the driver actually calls your upcall.

RaresCon replied:
This test is used just for the busy state of the driver.

alexandruradovici commented on apis/air_quality/src/tests.rs:
```diff
@@ -0,0 +1,106 @@
+}
+
+#[test]
+fn read_tvoc() {
```
Verify that the driver actually calls your upcall.

alexandruradovici reviewed (changes requested):
Please rebase.

alexandruradovici commented on apis/air_quality/src/lib.rs:
```diff
@@ -0,0 +1,92 @@
+        Self::read_data_sync(ReadTVOC)
+    }
+
+    fn read_data_sync(read_type: ReadType) -> Result<u32, ErrorCode> {
```
This is not similar to the async API, either make two functions either make one function for the synchronous API.

RaresCon replied:
As stated before, this function is not public, it has two wrappers that are public so the user of the API doesn't need to match the enum value for each command.

RaresCon committed:
Added suggested changes to AirQuality API

Added the suggested changes, including the `read_sync` function that
return a tuple of both CO2 and TVOC values. Improved the unittests for the
fake driver and added a unittest for the new function.

alexandruradovici reviewed (dismissed):
Please send this to upstream and post a link here.

RaresCon commented:
The PR is on upstream: https://github.com/tock/libtock-rs/pull/464

RaresCon committed:
Changed alphabetical order inside `Cargo.toml` and `lib.rs`

RaresCon committed:
Merge branch 'tock:master' into air_quality_api

RaresCon committed:
Added requested changes

I've added back the dedicated listener for this API
and rewritten the returning errors of some functions.

RaresCon committed:
Added documentation and fixed CI bug

I've added some documentation for the functions of the API
and fixed an inclusion bug of the API in `tests.rs` that
made the CI fail.

RaresCon committed:
Merge branch 'master' into air_quality_api
